python/tank/authentication/shotgun_shared/saml2_sso.py

```diff
@@ -0,0 +1,576 @@
+# Copyright (c) 2017 Autodesk.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+"""
+Module to support SSO login via a web browser and automated session renewal.
+
+This module, in conjunction with the slmodule, handles authentication on
+SSO enabled Shotgun websites. When we are in a session connected via SSO, we
+also handle automatic (and headless) renewal of the session.
+"""
+
+import base64
+import json
+import logging
+import os
+import time
+import urllib
+
+from .authentication_session_data import AuthenticationSessionData
+from ..ui.qt_abstraction import QtCore, QtNetwork, QtGui, QtWebKit
+from ... import LogManager
+
+log = LogManager.get_logger(__name__)
+
+
+# Error messages for events. . Also defined in slmodule/slutils.mu
+# @FIXME: Should import these from slmodule
+HTTP_CANT_CONNECT_TO_SHOTGUN = "Cannot Connect To Shotgun site."
+HTTP_AUTHENTICATE_SSO_NOT_UPPORTED = "SSO not supported or enabled on that site."
+HTTP_CANT_AUTHENTICATE_SSO_TIMEOUT = "Time out attempting to authenticate to SSO service."
+HTTP_CANT_AUTHENTICATE_SSO_NO_ACCESS = "You have not been granted access to the Shotgun site."
+
+# Paths for bootstrap the login/renewal process.
+URL_SAML_RENEW_PATH = "/saml/saml_renew"
+URL_SAML_RENEW_LANDING_PATH = "/saml/saml_renew_landing"
+
+# Old login path, which is not used for SSO.
+URL_LOGIN_PATH = "/user/login"
+
+
+class Saml2Sso(object):
+    """Performs Shotgun SSO login and pre-emptive renewal."""
+
+    def __init__(self, window_title="SSO"):
+        """Initialize the RV mode."""
+        log.debug("==- __init__")
+
+        self._event_data = None
+        self._sessions_stack = []
+        self._session_renewal_active = False
+
+        self._dialog = QtGui.QDialog()
+        self._dialog.setWindowTitle(window_title)
+        self._dialog.finished.connect(self.on_dialog_closed)
+
+        self._view = QtWebKit.QWebView(self._dialog)
```

Comment:
I need to do more tests on that.

In Maya 2017, I have verified that it does work.